### DISCUSSION TITLE
Pipeline with 2 NER: One trained, one from en_core_web_md

### DISCUSSION THREAD

**Discussion by oliviercwa** [explosion/spaCy#10340] [Help: Coding & Implementations]
Hello,
I am creating a pipeline with 2 NER:

1. the first one is Transformer-based and trained with my own dataset,
2. the second one is coming from Spacy en_core_web_md (component name = 'en-core-web-md-ner')

I am only training the first one but I want my final pipeline to have both: The trained transformer AND the en_core_web_md NER.

I don't have any problem with the first one, however for the second one, I need the model to be fully embedded as the en_core_web_md can't rely on the tensor created from the the first model.

Here is an excerpt from the config with the *en-core-web-md-ner* definition
```
[components.en-core-web-md-ner]
source = "en_core_web_md"
component = "ner"

frozen = ['en-core-web-md-ner']
```

However when I train my transformer, I am getting a UserWarning W113:
```
Sourced component 'en-core-web-md-ner' may not work as expected: source vectors are not identical to current pipeline vectors.
```

How can I tell the config to use en_core_web_md as source AND copy all embeddings (tok2vec) AND models from the source AND ignore the ones from the Transformer ?

Note #1 that I have also tried the following with the same outcome.
```
[components.en-core-web-md-ner]
source = "en_core_web_md"
component = "ner"

[components.en-core-web-md-ner.model]
@architectures = "spacy.TransitionBasedParser.v1"

[components.en-core-web-md-ner.model.tok2vec]
@architectures = "spacy.Tok2Vec.v2"

[components.en-core-web-md-ner.model.tok2vec.embed]
@architectures = "spacy.MultiHashEmbed.v2"

[components.en-core-web-md-ner.model.tok2vec.encode]
@architectures = "spacy.MaxoutWindowEncoder.v2"
```

Note #2: I have also tried to train first my transformer, load the resulting pipeline (my_pipeline) and add the en-core-web-md via APIs
```python
    source_nlp = spacy.load('en_core_web_md')
    component = my_pipeline.add_pipe('ner', source=source_nlp)
```
When I do this, I am getting the following warning
```
Sourced component 'ner' may not work as expected: source vectors are not identical to current pipeline vectors.
```

**Comment by adrianeboyd** [comment]
You want to source the component (with no additional settings beyond changing its name) and initialize the vectors from the same model:

```ini
[components.en-core-web-md-ner]
source = "en_core_web_md"
component = "ner"

[initialize]
vectors = "en_core_web_md"
```

A pipeline can only contain one set of vectors, but it's no problem to have both a `transformer` that doesn't use vectors and a `tok2vec` that does.

**Reply by oliviercwa** [reply]
Thanks Adriane. Regarding your comment 'A pipeline can only contain one set of vectors'. I understand the the doc has only one property vector. However are you saying that there is no way within the same pipeline to have:

- a transformer that uses its own embeddings (e.g. from HuggingFace transformer wrapper)
- AND a pretrained NER from "en_core_web_md" with its own embeddings ?

I was under the impression that by specifying in the configuration that the tok2vec model was not shared, the "en_core_web_md" NER component would "embed" them in its own layer, thus making it entirely independent from the rest of the pipeline. Essentially, ending up with 2 set of vectors:

- 1 set from the transformer that are also accessible using doc.vector
- 1 set "embedded" into the "en_core_web_md" NER model, which can't be accessed.

I guess I am missing something ?

Thanks

**Reply by adrianeboyd** [reply]
By "vectors" I mean the static word vectors that are stored in `nlp.vocab.vectors`. The `Tok2Vec` architecture uses these vectors with `include_static_vectors = true` while the `transformer` component doesn't. The vectors aren't saved within the `ner` component, which is why you have to add them separately with `[initialize.vectors]`. Since they're usually large and shared by multiple components, they're in one central place `nlp.vocab.vectors` rather than duplicated in each component.
